### ISSUE TITLE
Memory heap fragmentation

### ISSUE THREAD

**Issue by ardhipoetra** [ardhipoetra/faasm#4]
The backend (wavm, wamr) requires a lot of HEAP. Apparently, the worker couldn't run comfortably under 16GB of `SCONE_HEAP`. 32GB of `SCONE_HEAP` gives us more space, but much more than necessary.

Moreover, run it long enough, then you get a fragmentation as the following : 
```
[15:31:27] [51] [D] Opened preopened fd at /                                                                                                                                                                       
[15:31:27] [51] [D] Opened preopened fd at .                                                                                                                                                                       
mmap(0, 16384, PROT_READ|PROT_WRITE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0x1027051000, available memory: 16653548 kB
[SCONE|WARN] src/syscall/mman_anon.c:135:print_out_of_memory_warning(): Memory request for 8 GiB must be rejected as protected heap memory is exhausted (16 GiB of 32 GiB available)!
.......
In-Enclave Heap Memory Layout:                                                                           
[0x101f992000-0x1027055000] 121612 kB used                                                               
[0x1027055000-0x1226e00000] 8386220 kB empty                                                             
[0x1226e00000-0x1227002000] 2056 kB used                                                                 
[0x1227002000-0x1227200000] 2040 kB empty                                                                
[0x1227200000-0x1627200000] 16777216 kB used                                                             
[0x1627200000-0x109f992000] 8265288 kB empty                                                             
mmap(0, 8592031744, PROT_NONE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0xfffffffffffffff4, available memory: 16653548 kB
[SCONE|WARN] src/syscall/mman_anon.c:135:print_out_of_memory_warning(): Memory request for 8 GiB must be rejected as protected heap memory is exhausted (16 GiB of 32 GiB available)!
In-Enclave Heap Memory Layout:                                                                           
[0x101f992000-0x1027055000] 121612 kB used                                                               
[0x1027055000-0x1226e00000] 8386220 kB empty                                                             
[0x1226e00000-0x1227002000] 2056 kB used                                                                 
[0x1227002000-0x1227200000] 2040 kB empty
[0x1227200000-0x1627200000] 16777216 kB used
[0x1627200000-0x109f992000] 8265288 kB empty
mmap(0, 8592031744, PROT_NONE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0xfffffffffffffff4, available memory: 16653548 kB
[SCONE|WARN] src/syscall/mman_anon.c:135:print_out_of_memory_warning(): Memory request for 8 GiB must be rejected as protected heap memory is exhausted (16 GiB of 32 GiB available)!
In-Enclave Heap Memory Layout:
[0x101f992000-0x1027055000] 121612 kB used
[0x1027055000-0x1226e00000] 8386220 kB empty
[0x1226e00000-0x1227002000] 2056 kB used
[0x1227002000-0x1227200000] 2040 kB empty
[0x1227200000-0x1627200000] 16777216 kB used
[0x1627200000-0x109f992000] 8265288 kB empty
mmap(0, 8592031744, PROT_NONE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0xfffffffffffffff4, available memory: 16653548 kB
[SCONE|WARN] src/syscall/mman_anon.c:135:print_out_of_memory_warning(): Memory request for 8 GiB must be rejected as protected heap memory is exhausted (16 GiB of 32 GiB available)!
In-Enclave Heap Memory Layout:
[0x101f992000-0x1027055000] 121612 kB used
[0x1027055000-0x1226e00000] 8386220 kB empty
[0x1226e00000-0x1227002000] 2056 kB used
[0x1227002000-0x1227200000] 2040 kB empty
[0x1227200000-0x1627200000] 16777216 kB used
[0x1627200000-0x109f992000] 8265288 kB empty
mmap(0, 8592031744, PROT_NONE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0xfffffffffffffff4, available memory: 16653548 kB
[SCONE|WARN] src/syscall/mman_anon.c:135:print_out_of_memory_warning(): Memory request for 8 GiB must be rejected as protected heap memory is exhausted (16 GiB of 32 GiB available)!
In-Enclave Heap Memory Layout:
[0x101f992000-0x1027055000] 121612 kB used
[0x1027055000-0x1226e00000] 8386220 kB empty
[0x1226e00000-0x1227002000] 2056 kB used
[0x1227002000-0x1227200000] 2040 kB empty
[0x1227200000-0x1627200000] 16777216 kB used
[0x1627200000-0x109f992000] 8265288 kB empty
mmap(0, 8592031744, PROT_NONE, MAP_ANON|MAP_PRIVATE, -1, 0) = 0xfffffffffffffff4, available memory: 16653548 kB
munmap(0x1027051000, 16384) = 0, available memory: 16653564 kB
```

**Comment by ardhipoetra** [comment]
Created the issue on SCONE upstream : #2052

**Comment by ardhipoetra** [comment]
So far, there's nothing that can be done from the SCONE's perspective, since this issue is related to the memory allocation. 

AFAIK, wasm [memory model](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/memory_tune.md) need continuous memory allocation, hence the large, continuous allocation from the get-go.

Question for @csegarragonz :
- Is there anything configurable to tackle this?  Is there something page-table-like that might be possible? I tried `global heap` option as in [here](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/memory_tune.md) but it doesn't work as I imagined.
- How's wamr (or faasm) handling memory limitation on a hardware that has low amount of memory? Perhaps I can take some inspiration from that.
- What's your view if we're using the GC-enabled branch [here](https://github.com/bytecodealliance/wasm-micro-runtime/issues/2144)?
- on `./wamr/WAMRWasmModule.cpp`, it seems to call `bindInternal` on both loading and resetting wasm. `bindInternal` allocates memory which make sense when loading, but why on resetting too?

**Comment by ardhipoetra** [comment]
Disabling `hw_bound_check` (set `WAMR_DISABLE_HW_BOUND_CHECK` to 1), forcing WAMR to use `malloc` with only necessary memory used. By enabling HW-bound, it always map 8GB ([ref](https://github.com/bytecodealliance/wasm-micro-runtime/blob/0b2313f6f8fccf4ba89b553ced2987ac4b1d1230/core/iwasm/aot/aot_runtime.c#L388)).

However, by disabling that, we need to remove `wasm_runtime_set_exec_env_tls()` in [WAMRWasmModule.cpp:331](https://github.com/ardhipoetra/faasm/blob/892ea404e01a68c49a9926acff6c3ac911b48e1d/src/wamr/WAMRWasmModule.cpp#L331).

Random segfault after some time is still happening, though. (This is difficult to track, somehow)

Any input @csegarragonz ?

**Comment by csegarragonz** [comment]
Hey @ardhipoetra 

The 8 GB `mmap`-ing is how most WASM runtimes implement efficient bound checks.
Note that this is virtual memory, and is rarely used.
In summary, by `mmap`-ing 8 GB and placing the linear memory at the begining, the OS catch any out-of-bounds access marking the pages with the appropriate protections.
This, however, will be a problem inside an SGX enclave. (There is an issue covering it somewhere in the WAMR repo)

Note that WAMR already has an option to run WASM modules inside SGX enclaves (you can see how we enable it in Faasm using the env. variable `WASM_VM=sgx`). You may want to check what kind of flags we use to build WAMR there: https://github.com/faasm/faasm/blob/main/src/enclave/CMakeLists.txt

**Comment by ardhipoetra** [comment]
For the initial version, I disabled `hw_bound_check`. Some simple workload can be run without worrying about the heap fragmentation. I assume faasm doesn't have any hard requirement on this (except one usage below). 

However, I need to disable a call to `wasm_runtime_set_exec_env_tls` on `executeCatchException` as hinted by commit 132fa1225b1df62c0aed916cb126e59394693af8. Since it is related to exception, I guess it's pretty safe to put it aside for now.

Since that was disabled, a segfault is occurred when handling exception. Therefore, I changed one of the exception handling on https://github.com/ardhipoetra/faabric/commit/25fec0f0a330f4816f1a4be6e1205e658a7d885d. 

Both are band-aid patch and need to properly addressed when I understand what is going on.

Closing the issue for now.
